### ISSUE TITLE
Clarity on differences between 32-bit atomic extensions and core functions

### ISSUE THREAD

**Issue by oddhack** [KhronosGroup/OpenCL-Docs#1013]
The [extensions spec](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_int32_atomics) describes the functions provided by the extensions as

> These extensions became core features in OpenCL 1.1, except the built-in atomic function names are changed to use the atomic_ prefix instead of atom_ and the volatile qualifier was added to the pointer parameter p.

However, the OpenCL C spec isn't entirely consistent with this - it describes both atomic_ and atom_ functions, and all of them (with one exception which looks like an editing inconsistency) have the volatile qualifier. To complete the extension integration with the C spec, I'd like to figure out:

- Does the core provide *only* the atomic_ variants, or both atomic_ and atom_ variants of these functions?
- Do the atom_ variants as provided by the extensions indeed *not* have the volatile qualifier?

**Comment by bashbaug** [comment]
There is some additional information here, and in the referenced internal issue: https://github.com/KhronosGroup/OpenCL-Docs/issues/322

To answer your specific questions:

1. Yes, I believe that core OpenCL C only provides the functions prefixed by `atomic_`, whereas the extension provides the functions prefxed by `atom_`.  I really don't think this difference will be observable though, since we also say in the description of [CL_DEVICE_EXTENSIONS](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_EXTENSIONS) that "The following Khronos extension names must be returned by all devices that support OpenCL 1.1: [...] **cl_khr_global_int32_base_atomics** [...]", so OpenCL C 1.1 and newer effectively provides both versions.

3. It's possible that that some older implementations may have omitted the volatile qualifier, but all of the implementations we've surveyed now include the volatile qualifier, and including it seems to be the right thing to do, so I believe we've called the older specification without the volatile qualifier a spec bug, now fixed.  We should probably update (remove) the text saying that "the volatile qualifier was added to the pointer parameter p" accordingly.
